porch/pkg/engine/engine.go

```diff
@@ -846,86 +845,107 @@ func (cad *cadEngine) DeletePackage(ctx context.Context, repositoryObj *configap
+func (cad *cadEngine) UpdatePackageResources(ctx context.Context, repositoryObj *configapi.Repository, oldPackage *PackageRevision, old, new *api.PackageRevisionResources) (*PackageRevision, *api.RenderStatus, error) {
 	ctx, span := tracer.Start(ctx, "cadEngine::UpdatePackageResources", trace.WithAttributes())
 	defer span.End()
 
 	rev, err := oldPackage.repoPackageRevision.GetPackageRevision(ctx)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	// Validate package lifecycle. Can only update a draft.
 	switch lifecycle := rev.Spec.Lifecycle; lifecycle {
 	default:
-		return nil, fmt.Errorf("invalid original lifecycle value: %q", lifecycle)
+		return nil, nil, fmt.Errorf("invalid original lifecycle value: %q", lifecycle)
 	case api.PackageRevisionLifecycleDraft:
 		// Only draf can be updated.
 	case api.PackageRevisionLifecycleProposed, api.PackageRevisionLifecyclePublished:
 		// TODO: generate errors that can be translated to correct HTTP responses
-		return nil, fmt.Errorf("cannot update a package revision with lifecycle value %q; package must be Draft", lifecycle)
+		return nil, nil, fmt.Errorf("cannot update a package revision with lifecycle value %q; package must be Draft", lifecycle)
 	}
 
 	repo, err := cad.cache.OpenRepository(ctx, repositoryObj)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-
 	draft, err := repo.UpdatePackageRevision(ctx, oldPackage.repoPackageRevision)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	mutations := []mutation{
 		&mutationReplaceResources{
 			newResources: new,
 			oldResources: old,
 		},
-		&renderPackageMutation{
-			renderer: cad.renderer,
-			runtime:  cad.runtime,
-		},
 	}
-
-	apiResources, err := oldPackage.repoPackageRevision.GetResources(ctx)
+	prevResources, err := oldPackage.repoPackageRevision.GetResources(ctx)
 	if err != nil {
-		return nil, fmt.Errorf("cannot get package resources: %w", err)
+		return nil, nil, fmt.Errorf("cannot get package resources: %w", err)
 	}
 	resources := repository.PackageResources{
-		Contents: apiResources.Spec.Resources,
+		Contents: prevResources.Spec.Resources,
+	}
+	appliedResources, _, err := applyResourceMutations(ctx, draft, resources, mutations)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	if err := applyResourceMutations(ctx, draft, resources, mutations); err != nil {
-		return nil, err
+	// render the package
+	_, renderStatus, err := applyResourceMutations(ctx,
+		draft,
+		appliedResources,
+		[]mutation{&renderPackageMutation{
+			renderer: cad.renderer,
+			runtime:  cad.runtime,
+		}})
+	if err != nil {
+		// Render failure will not fail the overall API operation.
+		// The render error and result is captured as part of renderStatus above
+		// and is returned in packageresourceresources API's status field. We continue with
+		// saving the non-rendered resources to avoid losing user's changes.
+		// and supress this err.
+		err = nil
```

Comment:
Do we need this block? If we don't use the `err` variable for anything, can't we just replace it with `_` in the call to `applyResourceMutations`?

Reply:
Good suggestion. Done. Moved the comment above it.